File: church/church/report/department_strength/department_strength.js
// Copyright (c) 2026, kunle and contributors
// For license information, please see license.txt
/* eslint-disable */

// church/church/report/department_strength/department_strength.js

frappe.query_reports["Department Strength"] = {

    filters: [
        {
            fieldname: "department",
            label: __("Department"),
            fieldtype: "Link",
            options: "Church Department"
        },
        {
            fieldname: "show_members",
            label: __("Show Member Detail"),
            fieldtype: "Check",
            default: 0,
            description: "Tick to expand each department and show individual members"
        }
    ],

    formatter: function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        if (!data) return value;

        const is_dept_row   = !!data.total_members;   // department summary row
        const is_member_row = !!data.member_id;       // individual member row

        // Department name — bold, coloured header row
        if (column.fieldname === "department") {
            if (is_dept_row && data.department) {
                value = `<strong style="color:#1F3864;font-size:13px;">
                             🏢 ${data.department}
                         </strong>`;
            } else if (is_member_row) {
                value = "";  // no dept name on member rows
            }
        }

        // Active / inactive count
        if (column.fieldname === "active_members" && is_dept_row) {
            const n = parseInt(data.active_members) || 0;
            value = `<span style="color:#27ae60;font-weight:700;">${n}</span>`;
        }
        if (column.fieldname === "inactive_members" && is_dept_row) {
            const n = parseInt(data.inactive_members) || 0;
            value = n > 0
                ? `<span style="color:#e74c3c;">${n}</span>`
                : `<span style="color:#95a5a6;">0</span>`;
        }

        // Total members — large bold
        if (column.fieldname === "total_members" && is_dept_row) {
            const n = parseInt(data.total_members) || 0;
            const col = n >= 20 ? "#27ae60" : n >= 10 ? "#f39c12" : "#3498db";
            value = `<strong style="color:${col};font-size:14px;">${n}</strong>`;
        }

        // Member row — full name link
        if (column.fieldname === "full_name" && is_member_row) {
            value = `<span style="padding-left:20px;color:#2c3e50;">
                         ${data.full_name || ""}
                     </span>`;
        }

        // Member row — is_primary badge
        if (column.fieldname === "is_primary" && is_member_row) {
            value = data.is_primary
                ? `<span style="background:#667eea;color:#fff;
                                 padding:2px 6px;border-radius:8px;font-size:10px;">
                       ⭐ Primary
                   </span>`
                : "";
        }

        // Gender split — male/female bar
        if (column.fieldname === "male_count" && is_dept_row) {
            const m = parseInt(data.male_count)   || 0;
            const f = parseInt(data.female_count) || 0;
            const total = m + f || 1;
            const mpct  = Math.round(m / total * 100);
            value = `<div style="display:flex;align-items:center;gap:4px;">
                         <span style="color:#3498db;font-size:11px;">♂</span>
                         <div style="flex:1;background:#ecf0f1;border-radius:3px;height:8px;min-width:40px;">
                             <div style="width:${mpct}%;background:#3498db;height:8px;border-radius:3px;"></div>
                         </div>
                         <span style="font-size:11px;color:#3498db;">${m}</span>
                     </div>`;
        }
        if (column.fieldname === "female_count" && is_dept_row) {
            const f = parseInt(data.female_count) || 0;
            const m = parseInt(data.male_count)   || 0;
            const total = m + f || 1;
            const fpct  = Math.round(f / total * 100);
            value = `<div style="display:flex;align-items:center;gap:4px;">
                         <span style="color:#e91e8c;font-size:11px;">♀</span>
                         <div style="flex:1;background:#ecf0f1;border-radius:3px;height:8px;min-width:40px;">
                             <div style="width:${fpct}%;background:#e91e8c;height:8px;border-radius:3px;"></div>
                         </div>
                         <span style="font-size:11px;color:#e91e8c;">${f}</span>
                     </div>`;
        }

        return value;
    },

    // Alternate background for department summary rows vs member rows
    get_datatable_options: function(options) {
        return options;
    },

    onload: function(report) {
        report.page.add_inner_button(__("📋 Summary Only"), function() {
            frappe.query_report.set_filter_value("show_members", 0);
            frappe.query_report.refresh();
        });

        report.page.add_inner_button(__("👥 Show All Members"), function() {
            frappe.query_report.set_filter_value("show_members", 1);
            frappe.query_report.refresh();
        });
    }
};

File: church/church/report/department_strength/department_strength.py
# Copyright (c) 2026, kunle and contributors
# For license information, please see license.txt

# church/church/report/department_strength/department_strength.py

import frappe
from frappe import _

def execute(filters=None):
    filters = filters or {}
    columns = get_columns(filters)
    data    = get_data(filters)
    summary = get_summary(data)
    chart   = get_chart(data)
    return columns, data, None, chart, summary

def get_columns(filters):
    cols = [
        {"fieldname": "department",      "label": _("Department"),      "fieldtype": "Link", "options": "Church Department", "width": 180},
        {"fieldname": "total_members",   "label": _("Total Members"),   "fieldtype": "Int",  "width": 120},
        {"fieldname": "active_members",  "label": _("Active"),          "fieldtype": "Int",  "width": 90},
        {"fieldname": "inactive_members","label": _("Inactive"),        "fieldtype": "Int",  "width": 90},
        {"fieldname": "primary_count",   "label": _("Primary Dept of"), "fieldtype": "Int",  "width": 130},
        {"fieldname": "male_count",      "label": _("Male"),            "fieldtype": "Int",  "width": 70},
        {"fieldname": "female_count",    "label": _("Female"),          "fieldtype": "Int",  "width": 80},
        {"fieldname": "avg_age",         "label": _("Avg Age"),         "fieldtype": "Float","width": 90},
    ]
    if filters.get("show_members"):
        cols += [
            {"fieldname": "member_id",  "label": _("Member ID"),   "fieldtype": "Link", "options": "Member", "width": 130},
            {"fieldname": "full_name",  "label": _("Full Name"),   "fieldtype": "Data", "width": 180},
            {"fieldname": "role",       "label": _("Role"),        "fieldtype": "Data", "width": 140},
            {"fieldname": "is_primary", "label": _("Is Primary"),  "fieldtype": "Check","width": 100},
            {"fieldname": "from_date",  "label": _("Since"),       "fieldtype": "Date", "width": 100},
        ]
    return cols

def get_data(filters):
    dept_filter = "AND md.department = %(department)s" if filters.get("department") else ""

    # Department-level summary
    dept_rows = frappe.db.sql(f"""
        SELECT
            md.department,
            COUNT(md.name)                                        AS total_members,
            SUM(CASE WHEN md.is_active = 1 THEN 1 ELSE 0 END)    AS active_members,
            SUM(CASE WHEN md.is_active = 0 THEN 1 ELSE 0 END)    AS inactive_members,
            SUM(CASE WHEN md.is_primary = 1 THEN 1 ELSE 0 END)   AS primary_count,
            SUM(CASE WHEN m.gender = 'Male'   THEN 1 ELSE 0 END) AS male_count,
            SUM(CASE WHEN m.gender = 'Female' THEN 1 ELSE 0 END) AS female_count,
            ROUND(AVG(m.age), 1)                                  AS avg_age
        FROM `tabMember Department` md
        JOIN `tabMember` m ON m.name = md.parent
        WHERE m.member_status = 'Active'
          {dept_filter}
        GROUP BY md.department
        ORDER BY active_members DESC
    """, {"department": filters.get("department", "")}, as_dict=True)

    if not filters.get("show_members"):
        return dept_rows

    # Detailed member rows
    result = []
    for dept in dept_rows:
        result.append(dept)  # Department summary row
        members = frappe.db.sql("""
            SELECT
                md.parent    AS member_id,
                m.full_name,
                md.role,
                md.is_primary,
                md.from_date
            FROM `tabMember Department` md
            JOIN `tabMember` m ON m.name = md.parent
            WHERE md.department = %(dept)s
              AND m.member_status = 'Active'
              AND md.is_active = 1
            ORDER BY m.full_name
        """, {"dept": dept.department}, as_dict=True)
        for mem in members:
            result.append({
                "department": "",
                "member_id":  mem.member_id,
                "full_name":  mem.full_name,
                "role":       mem.role,
                "is_primary": mem.is_primary,
                "from_date":  mem.from_date,
            })
    return result

def get_summary(data):
    dept_rows  = [d for d in data if d.get("total_members")]
    total_dept = len(dept_rows)
    total_mem  = sum(d.active_members for d in dept_rows)
    largest    = max(dept_rows, key=lambda d: d.active_members) if dept_rows else None
    multi_dept = frappe.db.sql("""
        SELECT COUNT(DISTINCT parent) AS cnt
        FROM `tabMember Department`
        WHERE is_active = 1
        GROUP BY parent
        HAVING COUNT(name) > 1
    """, as_dict=True)
    multi_count = len(multi_dept)
    return [
        {"value": total_dept,  "label": _("Active Departments"),       "datatype": "Int",  "indicator": "blue"},
        {"value": total_mem,   "label": _("Total Dept Assignments"),   "datatype": "Int",  "indicator": "green"},
        {"value": multi_count, "label": _("Members in 2+ Depts"),      "datatype": "Int",  "indicator": "orange"},
        {"value": largest.department if largest else "—",
                               "label": _("Largest Department"),       "datatype": "Data", "indicator": "blue"},
    ]

def get_chart(data):
    dept_rows = [d for d in data if d.get("total_members")][:12]
    return {
        "data": {
            "labels": [d.department for d in dept_rows],
            "datasets": [
                {"name": _("Active"),   "values": [d.active_members   for d in dept_rows], "chartType": "bar"},
                {"name": _("Inactive"), "values": [d.inactive_members for d in dept_rows], "chartType": "bar"},
            ]
        },
        "type": "bar",
        "title": _("Department Strength"),
        "colors": ["#2ecc71","#e74c3c"],
        "barOptions": {"stacked": 1},
        "axisOptions": {"xIsSeries": 1}
    }
